### PR TITLE
Refactor CopyMoveFileHandlerTest as of #4900

### LINKED ISSUE #4923: Follow #4900 for CopyMovefileHandlerTest
Parent Issue: https://github.com/kiwix/kiwix-android/issues/4900

### PR DESCRIPTION
**Pull request by atharvyadav22**
Fixes: #4923 

- This is the last and the biggest test PR now all `flakyTest{}` wrapper instances are deleted.
- The original tests heavily relied on `relaxed = true` which silenced many tests errors.
- Originally there were **27 cases** indeterministic and redundant now there are now **45 tests cases** target each branch properly with behaviour test too.
- Added Main Dispatcher Rule for Dispatcher.Main controlled behaviour.

### CODE REVIEW

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz @gouri-panda 
Have a look at the PR description.
Thanks

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt [hunk 1]
```diff
@@ -332,7 +339,7 @@ class CopyMoveFileHandler @Inject constructor(
         destinationFile = destinationFile,
         copyMoveProgressBarController::updateProgress
       )
-      withContext(Dispatchers.Main) {
+      withContext(dispatcher) {
         if (moveSuccess) {
           notifyFileOperationSuccess(destinationFile, requireSelectedFileUri())
```
Comment (gouri-panda) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt:337]: @atharvyadav22 This is a regression. Previously, we did the UI task in the main thread now we are doing in the IO thread.

**Review by gouri-panda — COMMENTED** [review]
@atharvyadav22 This PR is still in draft so i'm assuming you're still working on it! I've found a lot of UI thread work done in the IO thread please remove these well. Also, the PR title is a bit misleading 🤔 ?

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt [hunk 2]
```diff
@@ -263,13 +270,13 @@ class CopyMoveFileHandler @Inject constructor(
   }
 
   private fun onCopyClicked(showStorageSelectionDialog: Boolean) {
-    lifecycleScope?.launch {
+    lifecycleScope?.launch(dispatcher) {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt]: This should not be performed on the `IO` thread. It should be performed on the `Main` thread. Inside this method, there are many UI related tasks.


@atharvyadav22 If you need a dispatcher with injection, you should create a new dispatcher with DI. Like we have the `@IoDispatcher`. You can create the `@MainDispatcher` so that you can pass it from test cases to avoid flakiness due to coroutine switches.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt [hunk 3]
```diff
@@ -263,13 +270,13 @@ class CopyMoveFileHandler @Inject constructor(
...
       performCopyOperation(showStorageSelectionDialog)
     }
   }
 
   private fun onMoveClicked(showStorageSelectionDialog: Boolean) {
-    lifecycleScope?.launch {
+    lifecycleScope?.launch(dispatcher) {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt]: Same for this.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt [hunk 4]
```diff
@@ -309,7 +316,7 @@ class CopyMoveFileHandler @Inject constructor(
         destinationFile,
         copyMoveProgressBarController::updateProgress
       )
-      withContext(Dispatchers.Main) {
+      withContext(dispatcher) {
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt]: It should use the `Main` thread instead of the `IO` thread.

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt (hunk 1 above)
Comment (MohitMaliFtechiz) [reply app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt:337]: Agree with @gouri-panda .

Comment (atharvyadav22) [reply app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt:337]: Yes, for this reason its an draft actually we have mixed UI and IO work in this file i am planning of changing it.

For this i requested an review for test cases are these fine? according to PR description.

**Review by MohitMaliFtechiz — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt [hunk 5]
```diff
@@ -84,6 +86,11 @@ class CopyMoveFileHandler @Inject constructor(
     this.unitTestDestinationFile = unitTestDestinationFile
   }
 
+  @VisibleForTesting
+  fun setStorageObservingJob(job: Job?) {
+    storageObservingJob = job
```
Comment (MohitMaliFtechiz) [inline app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt]: We are setting this explicitly and testing that. But it should test the actual one that is created by the code.

File: app/src/test/java/org/kiwix/kiwixmobile/localLibrary/CopyMoveFileHandlerTest.kt [hunk 6]
```diff
@@ -33,594 +29,1031 @@ import io.mockk.coVerify
...
-  fun `DetectingFileSystem with file less than 4GB continues operation`() = flakyTest {
-    runTest {
+  @Nested
+  inner class ShowMoveFileToPublicDirectoryDialog {
+    @Test
+    fun getStorageDeviceList_whenIsEmpty_showPreparingCopyMoveDialog() = runTest {
       fileHandler = spyk(fileHandler)
-      coEvery { storageCalculator.availableBytes(storageFile) } returns 10_000L
-      every { fat32Checker.fileSystemStates.value } returns DetectingFileSystem
+
```
Comment (MohitMaliFtechiz) [inline app/src/test/java/org/kiwix/kiwixmobile/localLibrary/CopyMoveFileHandlerTest.kt:121]: We should test the moveOperation as well. There are test cases written for the copy operation, but for the move operation there are no tests.

**Review by MohitMaliFtechiz — CHANGES_REQUESTED** [review]

File: app/src/test/java/org/kiwix/kiwixmobile/localLibrary/CopyMoveFileHandlerTest.kt (hunk 6 above)
Comment (atharvyadav22) [reply app/src/test/java/org/kiwix/kiwixmobile/localLibrary/CopyMoveFileHandlerTest.kt:121]: I have already added them
`inner class PerformMoveOperation {} `
here

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt (hunk 5 above)
Comment (atharvyadav22) [reply app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt]: i have added all the cases under `inner class Dispose {}` the reason i added this fun is because `storageObservingJob` is private

File: app/src/test/java/org/kiwix/kiwixmobile/localLibrary/CopyMoveFileHandlerTest.kt (hunk 6 above)
Comment (MohitMaliFtechiz) [reply app/src/test/java/org/kiwix/kiwixmobile/localLibrary/CopyMoveFileHandlerTest.kt:121]: image

The method is fully untested.

Also, please add unit test cases for `getDestinationFile`(with edge cases) and remove the `testGetDestinationFile` UI test.

We should verify that the `fileOperationHandler.move` method is invoked when `move` happens. Then we can ensure that moving is happening. We are only verifying `onMultipleFilesProcessSelection` is called. It does not fully test the functionality. We should add test cases which acutally tested the real behaviour.

```
fileCopyMoveCallback.onMultipleFilesProcessSelection(
          MultipleFilesProcessAction.Move
        )
```

Comment (atharvyadav22) [reply app/src/test/java/org/kiwix/kiwixmobile/localLibrary/CopyMoveFileHandlerTest.kt:121]: >We should test the moveOperation as well. There are test cases written for the copy operation, but for the move operation there are no tests.

Okay got it i thought `performMove()` I'll add

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt (hunk 2 above)
Comment (atharvyadav22) [reply app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt]: okay got it I'll create

File: app/src/test/java/org/kiwix/kiwixmobile/localLibrary/CopyMoveFileHandlerTest.kt (hunk 6 above)
Comment (atharvyadav22) [reply app/src/test/java/org/kiwix/kiwixmobile/localLibrary/CopyMoveFileHandlerTest.kt:121]: @MohitMaliFtechiz 
since its private and to fully test ill need to use what methods call it or should i make it an internal and @VisibleForTesting then we can test it properly?
I feel 2nd method is more easy for testing.
What say??

Comment (MohitMaliFtechiz) [reply app/src/test/java/org/kiwix/kiwixmobile/localLibrary/CopyMoveFileHandlerTest.kt:121]: Use the methods that call this method for real testing. You can mock the behaviour and test each branch of this method.

**Comment by atharvyadav22** [comment]
@MohitMaliFtechiz @gouri-panda 
I have made all the improvements and update the description too.

File: app/src/androidTest/java/org/kiwix/kiwixmobile/localLibrary/CopyMoveFileHandlerTest.kt [hunk 7]
```diff
@@ -307,10 +297,7 @@ class CopyMoveFileHandlerTest : BaseActivityTest() {
 
   private fun getSelectedFile(fileName: String): File {
     val loadFileStream =
-      CopyMoveFileHandlerTest::class.java.classLoader?.getResourceAsStream(fileName)
-    require(loadFileStream != null) {
```
Comment (MohitMaliFtechiz) [inline app/src/androidTest/java/org/kiwix/kiwixmobile/localLibrary/CopyMoveFileHandlerTest.kt]: Why do we remove this? 

Image

File: app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt (hunk 5 above)
Comment (MohitMaliFtechiz) [reply app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/CopyMoveFileHandler.kt]: @atharvyadav22 the production code does not expose it. So we should not expose it in test cases, just for coverage.

File: app/src/test/java/org/kiwix/kiwixmobile/localLibrary/CopyMoveFileHandlerTest.kt [hunk 8]
```diff
@@ -34,593 +30,1130 @@ import io.mockk.every
...
+    @Test
+    fun sufficientStorage_doesNotNotifyInsufficientStorage() = runTest {
+      coEvery {
+        storageCalculator.availableBytes(storageFile)
+      } returns 5000L
+
+      fileHandler.performCopyMoveOperationIfSufficientSpaceAvailable(storageFile)
+
+      verify(exactly = 0) {
```
Comment (MohitMaliFtechiz) [inline app/src/test/java/org/kiwix/kiwixmobile/localLibrary/CopyMoveFileHandlerTest.kt]: Instead of this, we should test it should performed the copyMove opreation.

**Review by MohitMaliFtechiz — CHANGES_REQUESTED** [review]

File: app/src/test/java/org/kiwix/kiwixmobile/localLibrary/CopyMoveFileHandlerTest.kt (hunk 8 above)
Comment (MohitMaliFtechiz) [reply app/src/test/java/org/kiwix/kiwixmobile/localLibrary/CopyMoveFileHandlerTest.kt]: @atharvyadav22 I have made the changes.

**Review by MohitMaliFtechiz — DISMISSED** [review]
@atharvyadav22 Thank you. Once CI is green. we will merge.

**Review by gouri-panda — DISMISSED** [review]
@atharvyadav22 Thanks! A little bit of conflicts and rebase then we're good to go :)

**Comment by kelson42** [comment]
sorry new conflicts....

**Comment by MohitMaliFtechiz** [comment]
@kelson42 Conflicts are resolved.
